chain-signatures/node/src/indexer_common.rs

```diff
@@ -0,0 +1,455 @@
+                &self.sender_string()?,
+                &self.path(),
+            )),
+        }
+    }
+}
+
+pub enum RespondBidirectionalEvent {
+    Solana(signet_program::RespondBidirectionalEvent),
+    Hydration(HydrationRespondBidirectionalEvent),
+}
+
+impl RespondBidirectionalEvent {
+    pub fn request_id(&self) -> [u8; 32] {
+        match self {
+            RespondBidirectionalEvent::Solana(event) => event.request_id,
+            RespondBidirectionalEvent::Hydration(event) => event.request_id,
+        }
+    }
+
+    pub fn responder(&self) -> [u8; 32] {
+        match self {
+            RespondBidirectionalEvent::Solana(event) => event.responder.to_bytes(),
+            RespondBidirectionalEvent::Hydration(event) => event.responder,
+        }
+    }
+
+    pub fn serialized_output(&self) -> Vec<u8> {
+        match self {
+            RespondBidirectionalEvent::Solana(event) => event.serialized_output.clone(),
+            RespondBidirectionalEvent::Hydration(event) => event.serialized_output.clone(),
+        }
+    }
+
+    pub fn signature(&self) -> Signature {
+        match self {
+            RespondBidirectionalEvent::Solana(event) => {
+                crate::indexer_sol::to_mpc_signature(event.signature.clone()).unwrap()
+            }
+            RespondBidirectionalEvent::Hydration(event) => event.signature.clone(),
+        }
+    }
+
+    pub fn source_chain(&self) -> Chain {
+        match self {
+            RespondBidirectionalEvent::Solana(_) => Chain::Solana,
+            RespondBidirectionalEvent::Hydration(_) => Chain::Hydration,
+        }
+    }
+}
+
+pub enum SignatureRespondedEvent {
+    Solana(signet_program::SignatureRespondedEvent),
+    Hydration(HydrationSignatureRespondedEvent),
+}
+
+impl SignatureRespondedEvent {
+    pub fn source_chain(&self) -> Chain {
+        match self {
+            SignatureRespondedEvent::Solana(_) => Chain::Solana,
+            SignatureRespondedEvent::Hydration(_) => Chain::Hydration,
+        }
+    }
+
+    pub fn request_id(&self) -> [u8; 32] {
+        match self {
+            SignatureRespondedEvent::Solana(event) => event.request_id,
+            SignatureRespondedEvent::Hydration(event) => event.request_id,
+        }
+    }
+
+    pub fn signature(&self) -> Signature {
+        match self {
+            SignatureRespondedEvent::Solana(event) => {
+                crate::indexer_sol::to_mpc_signature(event.signature.clone()).unwrap()
+            }
+            SignatureRespondedEvent::Hydration(event) => event.signature.clone(),
+        }
+    }
+}
+
+pub(crate) trait SignatureEventTrait {
+    fn generate_request_id(&self) -> [u8; 32];
+    fn generate_sign_request(
+        &self,
+        entropy: [u8; 32],
+        total_timeout: Duration,
+    ) -> anyhow::Result<IndexedSignRequest>;
+    fn source_chain(&self) -> Chain;
+    fn sender_string(&self) -> String;
+}
+
+pub(crate) trait SignatureEvent: SignatureEventTrait + std::fmt::Debug {}
```

Comment:
optional simplification: I don't think having two separate traits has a benefit here. I'd merge them into one.

```suggestion
pub(crate) trait SignatureEvent: std::fmt::Debug {
    fn generate_request_id(&self) -> [u8; 32];
    fn generate_sign_request(
        &self,
        entropy: [u8; 32],
        total_timeout: Duration,
    ) -> anyhow::Result<IndexedSignRequest>;
    fn source_chain(&self) -> Chain;
    fn sender_string(&self) -> String;
}
```

Then you could also remove all the blank `impl SignatureEvent for _ {}` lines.

Reply:
sounds good!